### PR TITLE
Feat/calendar

### PR DESCRIPTION
**Pull request by coltonpowershifter**
## Description

Created Calendar component currently required by Unified Checkout and MyTelus teams. This component is built on the React Dates library which has Moment.js as a peer dependency. Note that the component has been styled using Styled Components, so it has slightly different structure.

See #60 

## Checklist before submitting pull request

- Commits follow our [Developer Guide](https://github.com/telus/tds-community/blob/chore/template-hints/.github/CONTRIBUTING.md)
- [x] New code is unit tested
- [x] make sure visual and accessibility tests pass
- [x] make sure code builds

### CODE REVIEW

**Comment by jacobdeepak** [comment]
We needed a Calendar for My Telus - Appointment Tool and found that the the above mentioned component  is a date picker not a calendar. Internally the above mentioned contribution makes use of `SingleDatePicker` from airbnb `react-dates` where as the we need to use `DayPicker` from the same library. The properties and method invoked to override to get to desired behaviour, look and feel is different for our use case.

**Comment by jacobdeepak** [comment]
Unfortunately we wont be able to use the above component as an appointment Calendar.

**Comment by mujtabahasni** [comment]
I tried playing with sample code `<Calendar/>` component in the doc, but because `moment.js` library isn't in scope of the styleguide,  I cannot create dates or date functions to play with `<Calender/>` props.  Consider adding `moment.js` to the require list in `styleguide.config.js`

**Comment by mujtabahasni** [comment]
>  we need to use `DayPicker` from the same library.

Perhaps an `inline` property can be added to switch display components, but otherwise use the same props/api?

**Comment by bnadeem-telus** [comment]
Enrico is going to follow up with Fabio

File: packages/Calendar/Calendar.jsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+Calendar.propTypes = {
+  /** A unique identifier */
+  id: PropTypes.string.isRequired,
+
+  /** A Moment instance representing the currently selected date <em>i.e. moment() </em> */
+  date: momentPropTypes.momentObj,
+
+  /** Event triggered every time a new date is clicked on
```
Comment (theetrain) [inline packages/Calendar/Calendar.jsx]: Replace `clicked on` with `selected`

File: packages/Calendar/Calendar.jsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  @param {Moment} date The new date that was clicked */
```
Comment (theetrain) [inline packages/Calendar/Calendar.jsx]: Replace `clicked` with `selected`

File: packages/Calendar/Calendar.jsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  onDateChange: PropTypes.func,
+
+  /** A function determing whether a given date should be disabled
```
Comment (theetrain) [inline packages/Calendar/Calendar.jsx]: Typo in `determining`

File: packages/Calendar/package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "@tds/community-calendar",
```
Comment (theetrain) [inline packages/Calendar/package.json]: Please double check the component name with your designer, I believe it should be `@tds/community-calendar-picker`

File: packages/Calendar/package.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  "peerDependencies": {
+    "react": ">=15",
+    "react-dom": ">=15"
+  },
+  "dependencies": {
+    "moment": "^2.24.0",
+    "prop-types": "^15.6.2",
+    "react-dates": "^19.0.3",
+    "react-moment-proptypes": "^1.6.0",
+    "styled-components": "^4.1.3"
```
Comment (theetrain) [inline packages/Calendar/package.json]: Please move `styled-components` under `peerDependencies`
```diff
"peerDependencies": {
  "react": ">=15",
-  "react-dom": ">=15"
+  "react-dom": ">=15",
+  "styled-components": ">=4"
}
```

File: packages/Calendar/package.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  "bugs": {
+    "url": "https://github.com/telus/tds-community/issues"
+  },
+  "homepage": "http://tds.telus.com",
+  "peerDependencies": {
+    "react": ">=15",
+    "react-dom": ">=15"
+  },
+  "dependencies": {
```
Comment (theetrain) [inline packages/Calendar/package.json]: Please add `@tds/core-decorative-icon` as a dependency

```sh
npx lerna add @tds/core-decorative-icon --scope @tds/community-calendar-picker
```

File: packages/Calendar/styles.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
+import styled from 'styled-components'
+
+const theme = {
+  colors: {
```
Comment (jraff) [inline packages/Calendar/styles.js]: These colours can now be imported using `@tds/core-colours@1.1.1` - https://tds.telus.com/components/index.html#colours

File: packages/Calendar/styles.js [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    telusPurple: '#4B286D',
+    shark: '#2A2C2E',
+    shuttleGrey: '#54595F',
+    white: '#FFF',
+    gainsboro: '#D8D8D8',
+  },
+  fontWeights: {
```
Comment (jraff) [inline packages/Calendar/styles.js]: These font weights can be be imported using `@tds/shared-typography@1.2.1` - https://github.com/telus/tds-core/blob/master/packages/typography/typography.js

File: packages/Calendar/styles.js [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    thin: 300,
+    light: 400,
+    roman: 500,
+    medium: 700,
+  },
+  breakPoints: {
```
Comment (jraff) [inline packages/Calendar/styles.js]: These breakpoints can now be imported using `@tds/core-responsive@1.2.0` See the **media utility usage** section for a handy helper for generating media queries similar to the existing `sass-mq` usage.

**Comment by donnavitan** [comment]
Hello all,

I just wanted to capture all the discussion thus far:

### Reference
- http://airbnb.io/react-dates/?path=/story/sdp-calendar-props--single-month
- https://telus.invisionapp.com/d/#/console/13362811/356301617/preview


### Outstanding things to solve for
1. TDS-Calendar bundle size after build. TDS-Calendar is using `react-dates` which has bundle size of over 600kb. The main reason is that it is using heavy packages like `moment` and `lodash`.
2. Responsiveness for screensize 323X426, i have proposed above mentioned solution ( decreasing the day cell size for better visibility)
4. Keeping focus on month navigation button, when we switch months using keyboard. - `Needs discussion` (edited) 


### Explored thus far as per discussions with @harmeetsaimbhi @xingbo828 @brianlam1 
*Limitation 1*
The minimum mobile viewport width as per Telus standard is 320px.
We also have padding of 16px both sides for the full screen container, so we are left with 320-(16*2) = 288px.
So it’s impossible to render a component of width 312px in a 288px container.


*Limitation 2*
The point of origin of calendar popover depends upon the input find above.
It dynamically calculated the `left` css property as per the location of above text input and `we cannot override this behaviour .`
If we look at the padding of the text input, it is 16 + 16 = 32 pixels both sides.
So 320 - 64 = 256px is the space left to render a 312px calendar popover, which is not possible.

### Recommendations
1. Re: bundle size: no further progress? If so, please add comments.

2. Re: responsiveness
- In order to fit within the prescribed small viewport (320px), the cell date width can be adjusted as long as the height can be kept for 40px
- Additionally, since this solution is to address the very edge of 320px limit, is it possible to add another breakpoint (max 344px) where the prescribed design solution would actually fit, rather than wait for the next breakpoint (it's a really big gap to serve a less ideal design experienced)
- Considerations for making this a fullscreen overlay (AKA modal) are not recommended due to the additional a11y requirements and development effort

3. Re: focus management
- Yes, please proceed with suggestion to keep focus on the button that was triggered instead of moving to the next item (cell date "1")


### Other considerations
- TDS highly recommends the use of native platform/device/browser date pickers whenever possible

**Comment by xingbo828** [comment]
1. Unfortunately the bundle size can't be further reduced as the moment  library is the core dependency of react-date package .
2. 40-48px is the minimum width of a touchable element.. are we sure we want to further reduce the width of cells? They are at 40px atm
3. Focus management is handled within react-date package, we can submit an issue to their repo but there is no guarantee when this can be fixed.

**Comment by donnavitan** [comment]
Hello all,

Re: Touch target
The recommendation to allow for the width to be flexible includes that the height must remain at 40px height; as well an additional breakpoint be added so that the cell date will become 40px width and height can be available as soon as possible without waiting for the next official TDS breakpoint.

To address known accessibility concerns, please note that the form input and native date picker are the preferred method. Since minimum target touch size recommendations can vary per platform/device. TDS will comply with WCAG AA, or better whenever possible. In which, the prescribed recommendation and still available input, works.

Re: Focus management
It's unfortunate that we don't have as much control over this tool. If the request can be made, that would be nice to have.

🌈🐸

**Comment by donnavitan** [comment]
Just wanted to add another update, since I had a chance to discuss with the accessibility team at their office hours:

Re: focus management
- Focus order can depend on the business goal. For example, if you were choosing an appointment date, maybe it is good to move focus to the next available date as a recommendation. In this regard, the business should be able to adjust focus order after navigation
- On the other hand, keeping focus on the navigation (next/previous month) where recommendation is not available, is also acceptable
- On the other, other hand, we don't have control over this tool at this time

So thurrr ya go!

**Comment by donnavitan** [comment]
As per discussion, I've added comments on an internal Jira ticket to review recommendations to implement a custom breakpoint to show optimum minimum width. @abansil will review and move forward with her preference.

For visual reference, this screenshot showcase 2 minimum width options:
1. 376px minimum width breakpoint with 40px dates
2. 432px minimum width breakpoint with 48px dates

image

**Comment by harmeetsaimbhi** [comment]
Created a new PR (https://github.com/telus/tds-community/pull/285) to avoid rebase 40 commits.

**Pull request closed without merging**